python/global_search_completer/global_search_completer.py

```diff
@@ -0,0 +1,362 @@
+        any QT UI components. It should not do any logging and should
+        be as minimal as possible.
+
+        :param sg: Shotgun instance
+        :param data: data dictionary passed in from _submit(). This dict
+            carries the options payload and is forwarded over by the
+            sg_data.execute_method() method. In our case, the dictionary
+            holds a single 'text' key which contains the search phrase.
+        """
+
+        # constrain by project in the search
+        project_ids = []
+        if self._bundle.context.project:
+            project_ids.append(self._bundle.context.project["id"])
+
+        # run the query
+        sg_data = sg.text_search(
+            data["text"],
+            self._entity_search_criteria,
+            project_ids
+        )
+
+        return sg_data
+
+    def __on_worker_failure(self, uid, msg):
+        """
+        Asynchronous callback - the worker thread errored.
+
+        :param uid: Unique id for request that failed
+        :param msg: Error message
+        """
+        uid = shotgun_model.sanitize_qt(uid) # qstring on pyqt, str on pyside
+        msg = shotgun_model.sanitize_qt(msg)
+        if self._processing_id == uid:
+            self._bundle.log_warning("Could not retrieve search results: %s" % msg)
+
+    def __on_worker_signal(self, uid, request_type, data):
+        """
+        Signaled whenever the worker completes something.
+        This method will dispatch the work to different methods
+        depending on what async task has completed.
+
+        :param uid: Unique id for request
+        :param request_type: String identifying the request class
+        :param data: the data that was returned
+        """
+        uid = shotgun_model.sanitize_qt(uid) # qstring on pyqt, str on pyside
+        data = shotgun_model.sanitize_qt(data)
+
+        if uid in self._thumb_map:
+            thumbnail = data["image"]
+            if thumbnail:
+                item = self._thumb_map[uid]["item"]
+                thumb = QtGui.QPixmap.fromImage(thumbnail)
+                item.setIcon(thumb)
+
+
+        if self._processing_id == uid:
+            # all done!
+            self._clear_model(add_loading_item=False)
+
+            matches = data["return_value"]["matches"]
+
+            if len(matches) == 0:
+                item = QtGui.QStandardItem("No matches found!")
+                item.setData(self.MODE_NOT_FOUND, self.MODE_ROLE)
+                self.model().appendRow(item)
+
+            # insert new data into model
+            for d in matches:
+                item = QtGui.QStandardItem(d["name"])
+                item.setData(self.MODE_RESULT, self.MODE_ROLE)
+
+                item.setData(shotgun_model.sanitize_for_qt_model(d),
+                             self.SG_DATA_ROLE)
+
+                item.setIcon(self._default_icon)
+
+                if d.get("image", None) and self.__sg_data_retriever:
```

Comment:
Is it possible not have __sg_data_retriever given that we're in the work completed callback here? Seems like it pretty much has to be available at this point.